system/lib/wasmfs/syscalls.cpp

```diff
@@ -1121,4 +1129,65 @@ long __syscall_ftruncate64(long fd, long low, long high) {
   }
   return ret;
 }
+
+long __syscall_pipe(long fd) {
+  auto* fds = (__wasi_fd_t*)fd;
+
+  // Make a pipe: Two PipeFiles that share a single data source between them, so
+  // that writing to one can be read in the other.
+  //
+  // No backend is needed here, so pass in nullptr for that.
+  auto data = std::make_shared<PipeData>();
+  auto reader = std::make_shared<PipeFile>(S_IRUGO, data);
+  auto writer = std::make_shared<PipeFile>(S_IWUGO, data);
+
+  auto fileTable = wasmFS.getFileTable().locked();
+  fds[0] =
+    fileTable.addEntry(std::make_shared<OpenFileState>(0, O_RDONLY, reader));
+  fds[1] =
+    fileTable.addEntry(std::make_shared<OpenFileState>(0, O_WRONLY, writer));
+
+  return 0;
+}
+
+int __syscall_poll(struct pollfd* fds, nfds_t nfds, int timeout) {
+  auto fileTable = wasmFS.getFileTable().locked();
+
+  // Process the list of FDs and compute their revents masks. Count the number
+  // of nonzero such masks, which is our return value.
+  long nonzero = 0;
+  for (nfds_t i = 0; i < nfds; i++) {
+    auto* pollfd = &fds[i];
+    auto fd = pollfd->fd;
+    if (fd < 0) {
+      // Negative FDs are ignored in poll().
+      continue;
+    }
+    // Assume invalid, unless there is an open file.
+    auto mask = POLLNVAL;
+    auto openFile = fileTable.getEntry(fd);
+    if (openFile) {
+      mask = 0;
+      auto flags = openFile->getFlags();
+      if (flags == O_WRONLY || flags == O_RDWR) {
+        mask = POLLOUT;
+      }
+      if (flags == O_RDONLY || flags == O_RDWR) {
+        // If there is data in the file, then there is also the ability to read.
+        if (openFile->locked().getFile()->locked().getSize() > 0) {
+          mask |= POLLIN;
+        }
+      }
+      // TODO: get mask from File dynamically using a poll() hook?
+    }
+    // Mask the relevant bits: the given events, and also the exceptional
+    // conditions of error and hangup.
+    mask &= pollfd->events | POLLERR | POLLHUP;
```

Comment:
Since this PR also adds pipes, it would be good to set `POLLERR` on the write end of the pipe when the read end is closed and to set `POLLHUP` on the read end of the pipe when the write end has been closed.

Reply:
What do you think about leaving those for followups? That is an increase in functionality compared to the old FS backend. Not that we need to match it exactly, but anything additional would not be tested yet. I'd rather add TODOs for things like this (added now).

Reply:
TODO sgtm.